### PR TITLE
Port FUSEPY to native Windows and Cygwin

### PR DESCRIPTION
**Pull request by billziss-gh**
The commits in this PR port FUSEPY to native Windows and Cygwin.

- The first commit defines the Cygwin `struct stat`. This allows FUSEPY to be used with Cygwin.
- The second commit also ports FUSEPY to native Windows.

This requires installation of [WinFsp](https://github.com/billziss-gh/winfsp) (disclaimer: I am the author). WinFsp exposes FUSE-like file system functionality on Windows. It has its own native API, but also provides a FUSE 2.8 API.

WinFsp supports most NTFS and POSIX features (including symlinks on Windows). Notable exceptions:
- Hard links which are supported by Windows, but not on WinFsp yet (there are plans to do so).
- Extended attributes are supported on Windows, but are severely limited compared to extended attributes on POSIX. They are not supported on WinFsp (this could easily change if a real use case for them emerges).

Caveats:
- FUSEPY on Windows has no issues that I am aware of.
- FUSEPY on Cygwin can only be safely run in single-threaded mode (`nothreads=True`), otherwise a segmentation fault may occur (with Python 2.7). I have researched this and the problem appears to be deep within Python, perhaps in the ctypes module.

For those interested my [fusepy fork](https://github.com/billziss-gh/fusepy) already has support for Windows.

### CODE REVIEW

**Comment by billziss-gh** [comment]
Ping

**Comment by billziss-gh** [comment]
I have added a commit to also port FUSEPY to native Windows and have updated the original description of this Pull Request. Thank you.

File: fuse.py [hunk 1]
```diff
@@ -46,6 +46,23 @@ def newfunc(*fargs, **fkeywords):
...
+    #
+    # sizeof(long)==4 on Windows 32-bit and 64-bit
+    # sizeof(long)==4 on Cygwin 32-bit and ==8 on Cygwin 64-bit
+    #
+    # We have to fix up c_long and c_ulong so that it matches the
+    # Cygwin (and UNIX) sizes when run on Windows. The alternative
+    # would be to modify every structure below that uses c_long and
+    # c_ulong.
+    import sys
+    c_long = c_int64 if sys.maxsize > 0xffffffff else c_int32
```
Comment (terencehonles) [inline fuse.py]: can you change this to one `if`

File: examples/memory.py [hunk 2]
```diff
@@ -22,7 +22,7 @@ def __init__(self):
         self.data = defaultdict(bytes)
         self.fd = 0
         now = time()
-        self.files['/'] = dict(st_mode=(S_IFDIR | 0o755), st_ctime=now,
+        self.files['/'] = dict(st_mode=(S_IFDIR | 0o777), st_ctime=now,
```
Comment (terencehonles) [inline examples/memory.py]: is this needed?

File: fuse.py [hunk 3]
```diff
@@ -55,13 +72,29 @@ class c_utimbuf(Structure):
...
     pass    # Platform dependent
 
-_system = system()
-_machine = machine()
-
 if _system == 'Darwin':
     _libiconv = CDLL(find_library('iconv'), RTLD_GLOBAL) # libfuse dependency
     _libfuse_path = (find_library('fuse4x') or find_library('osxfuse') or
                      find_library('fuse'))
+elif _system == 'Windows':
```
Comment (terencehonles) [inline fuse.py]: can we put this in a different PR? As of #31 there is the ENV var `FUSE_LIBRARY_PATH`

File: fuse.py (hunk 1 above)
Comment (billziss-gh) [reply fuse.py]: Will do.

File: examples/memory.py (hunk 2 above)
Comment (billziss-gh) [reply examples/memory.py]: I believe it was needed because the POSIX and Windows permission models are very different. Will retest and see if it is really necessary.

File: fuse.py (hunk 3 above)
Comment (billziss-gh) [reply fuse.py]: I have looked at the FUSE_LIBRARY_PATH changes. Perhaps we could handle this similarly to how Darwin is handled? See below:

```python
_libfuse_path = os.environ.get('FUSE_LIBRARY_PATH')
if not _libfuse_path:
    if _system == 'Darwin':
        # Darwin code
    elif _system == 'Windows':
        # Windows code
    else:
        _libfuse_path = find_library('fuse')
```

Unfortunately when I created this PR I created it off the `master` branch in my fork and there are now users that rely on that branch. This makes it hard to split the PR into two (at least with my limited GitHub skills). It might be easier to put everything in the same PR (unless you do not intend to incorporate this particular change into the main project).

**Comment by billziss-gh** [comment]
Thank you for your comments. Please see above for my responses.

File: examples/memory.py (hunk 2 above)
Comment (terencehonles) [reply examples/memory.py]: Thanks, let me know what you find. Otherwise I'll have to spin up a windows box.

File: fuse.py (hunk 3 above)
Comment (terencehonles) [reply fuse.py]: @billziss-gh I was mainly interested in merging in the changes sans registry code and wanted to start a separate discussion on the registry related changes.

If you're not too familiar with git I can probably look at squashing your changes and creating a new PR where @eestrada and others can discuss the change.

Comment (billziss-gh) [reply fuse.py]: Understood. I will create a new branch with just the Windows/Cygwin `stat` related changes (and without the registry code) and probably end up submitting as a new PR. I am currently very busy with multiple projects, but I am hoping to have this for you this or next week.

Comment (terencehonles) [reply fuse.py]: @billziss-gh thank you! :slightly_smiling_face:

**Comment by N-Coder** [comment]
I had some minor issues with the `loopback.py` example on Windows, because some functions are not available or work differently.

@billziss-gh if you could rebase your fork to the latest version from this repository (merging the 2 conflicting files and including the more recent changes to the loopback.py file as I don't want to cause additional merge conflicts) I'd be glad to make a PR commenting out the missing functions and fixing up the rest and giving my thumbs up for this PR. :)

**Comment by billziss-gh** [comment]
I will be working today to make this happen. Expect to see some activity on the master branch of my fork, which this PR pulls from. Because GitHub does not appear to let us change branches that a PR pulls from I may have to create a new PR. Hopefully this should happen in a couple of hours.

**Comment by billziss-gh** [comment]
BTW, in working through this I notice that a lot of struct definitions have been changed to use `ctypes.C_TYPE` rather than simply `C_TYPE`.

This interferes with this code, which is unfortunately necessary on Windows:

```python
if _system == 'Windows':
    # NOTE:
    #
    # sizeof(long)==4 on Windows 32-bit and 64-bit
    # sizeof(long)==4 on Cygwin 32-bit and ==8 on Cygwin 64-bit
    #
    # We have to fix up c_long and c_ulong so that it matches the
    # Cygwin (and UNIX) sizes when run on Windows. The alternative
    # would be to modify every structure below that uses c_long and
    # c_ulong.
    import sys
    if sys.maxsize > 0xffffffff:
        c_long = c_int64
        c_ulong = c_uint64
    else:
        c_long = c_int32
        c_ulong = c_uint32
```

I can of course revert the relevant uses of `ctypes.c_long` and `ctypes.c_ulong` in my PR, but I would prefer to hear from the maintainer(s).

**Comment by billziss-gh** [comment]
Please ignore my earlier comment. I have solved this problem in a different way.

Expect a new PR soon.

**Comment by billziss-gh** [comment]
I have now submitted new PR #107 and I am closing this one.
